### PR TITLE
raise if association doesn't have a primary key

### PR DESCRIPTION
**Pull request by greg-rychlewski**
Related to https://github.com/elixir-ecto/ecto/issues/3855

When an association doesn't have a primary key, the data structure used to track the relationships between associated structs doesn't work as expected because it relies on the primary key. This PR raises early if any of the associations are missing a primary key. 

The queries can still technically function correctly if the leaf association doesn't have a PK, but I'm assuming the desired behaviour is to raise for symmetry/consistency reasons. For instance, if the user switched the order of the query so that the leaf is the parent it would no longer work. But if I'm mistaken please let me know!

### CODE REVIEW

File: lib/ecto/repo/assoc.ex
Comment (josevalim) [inline lib/ecto/repo/assoc.ex:114]:
```suggestion
```

**Comment by josevalim** [comment]
Hi Greg! Thank you for the amazing for pull requests. Would you like to join the Ecto team?

It doesn’t Require new commitments From your side. But All help is appreciated And you have been doing a fantastic job. It is your decision, no pressure! just let me know :)

Have a great weekend!

**Comment by greg-rychlewski** [comment]
Hi @josevalim 

I would love to join the Ecto team! I feel extremely honoured that you asked. Thank you so much.

I put my email and location in my profile now, in case you or anyone else would like a more direct way to contact me. And so you know when I might be asleep.

Have a great weekend! Thanks again!

**Comment by josevalim** [comment]
Invite sent, welcome to the team!

Feel free to merge your own PRs from now on after they get at least one approval by another team member. Also feel free to merge PR from external contributors. :)

**Review by josevalim — APPROVED** [review]

**Comment by JustMikey** [comment]
Thanks for the PR but I am afraid it was a breaking change for me and I wonder if we should update the changelog file for the version `3.9.1`.

For example, in my case I have a schema `Profile` using a table `profiles` and a schema `ProfileStats` using a materialized view `profile_stats`, where the latter simply contains the analytics about the former and as such, it doesn't have a primary key, only column `profile_id` that is used as a foreign key, so the schema definition has `@primary_key false` and `belongs_to :profile, Profile` and it has stopped working after I updated to `3.9.1`. It used to work both with inner joins and preloads but now it seems to be broken. And I imagine other folks may end up in a similar situation after the upgrade.

**Comment by greg-rychlewski** [comment]
Thanks for letting us know @JustMikey. That sounds like a good idea to me. I think this one went into 3.8.0. 

@josevalim could I submit a modified changelog and have it published to hexdoc or is it too late?

**Comment by josevalim** [comment]
Let’s update the CHANGELOG and it will be fixed in the last release. :)

**Pull request merged**
